tsdb/head.go

```diff
@@ -2023,6 +2023,9 @@ type memSeries struct {
 	lset labels.Labels
 	meta *metadata.Metadata
 
+	// Series labels hash to use for sharding purposes.
+	shardHash uint64
```

Comment:
Keeping the hash here significantly improve performance. It's a must for Mimir. In Mimir we haven't measured a significant extra cost to keep it, and I don't know if it really makes sense to upstream this feature without it (cause we'll have to change the implementation again in mimir-prometheus to keep the current behaviour).

Reply:
We could use a build tag to have it in Mimir and omit it in Prometheus.
(e.g. `shardHash` is a field in a struct which is declared in a new file, and that struct is brought here as an anonymous member).

Reply:
Tag is too much maybe, but `memSeriesShardable` would work. I am fine to merge for now in this form and improve later on if we see regression 👍🏽

Reply:
> separate struct memSeriesShardable would work

I would like to better understand your idea. If we have a separate struct `memSeriesShardable` how does it reduces the memory footprint unless we use build tags with two different versions?

> I am fine to merge for now in this form

There are other comments from Bryan I would like to address first.

> improve later on if we see regression

What if we run the prombench on this PR?

Reply:
> Two different newMemSeries() implementations but same signature?

Yes, one that doesn't compute the hash.  Two implementations for most of the functions added in this PR.

Reply:
The prombench you've run shows no difference in memory, but an increased CPU usage. Given the build tag option here would have been useful to not impact the memory (at the cost of extra complexity), but memory looks not impacted anyway, WDYT about just skipping computing the shard adding a `EnableSharding` config option to TSDB?

My proposal is in this comment: https://github.com/prometheus/prometheus/pull/10421/commits/c38d4e494c117d1700fddd0d32cb78309ccce1e1

Reply:
More detail on the point I made earlier: Go allocates heap objects into a size class, e.g. it will allocate 224 or 240 or 256 bytes for this struct. If it currently needs 248 and you add 8 more then no difference is observed, however it is now twice as hard to drop it to the 240 class.

I'm ok leaving that to a future PR, but build tags favouring different consumers is something I would like to investigate.